### PR TITLE
Add AffineTransformation

### PR DESCRIPTION
**Pull request by ekagra-ranjan**
In reference to #569 added AffineTransformation to superseed LinearTransformation.

### CODE REVIEW

File: torchvision/transforms/transforms.py [hunk 1]
```diff
@@ -702,52 +702,68 @@ def __repr__(self):
...
 
+class LinearTransformation(object):
+    """
+    Note: This transform is deprecated in favor of AffineTransformation.
+    """
+
+    def __init__(self, transformation_matrix):
+        warnings.warn("The use of the transforms.LinearTransformation transform is deprecated, " +
+                      "please use transforms.AffineTransformation instead.")
+        super(LinearTransformation, self).__init__(transformation_matrix)
```
Comment (fmassa) [inline torchvision/transforms/transforms.py]: I suppose you want to pass a `mean_vector` filled with zeros here?

File: torchvision/transforms/transforms.py [hunk 2]
```diff
@@ -702,52 +702,68 @@ def __repr__(self):
...
+                             " as the transformation_matrix")
+
         self.transformation_matrix = transformation_matrix
+        self.mean_vector = mean_vector
 
     def __call__(self, tensor):
         """
         Args:
             tensor (Tensor): Tensor image of size (C, H, W) to be whitened.
-
```
Comment (fmassa) [inline torchvision/transforms/transforms.py]: Can you keep those spaces here and above? I think they are needed to better render the documentation

File: torchvision/transforms/transforms.py [hunk 3]
```diff
@@ -27,7 +27,7 @@
 __all__ = ["Compose", "ToTensor", "ToPILImage", "Normalize", "Resize", "Scale", "CenterCrop", "Pad",
            "Lambda", "RandomApply", "RandomChoice", "RandomOrder", "RandomCrop", "RandomHorizontalFlip",
            "RandomVerticalFlip", "RandomResizedCrop", "RandomSizedCrop", "FiveCrop", "TenCrop", "LinearTransformation",
-           "ColorJitter", "RandomRotation", "RandomAffine", "Grayscale", "RandomGrayscale"]
+           "AffineTransformation", "ColorJitter", "RandomRotation", "RandomAffine", "Grayscale", "RandomGrayscale"]
```
Comment (fmassa) [inline torchvision/transforms/transforms.py]: Don't you need to keep `LinearTransformation` here as well for it to be visible?

Comment (ekagra-ranjan) [reply torchvision/transforms/transforms.py]: It is mentioned in the last part of line 29.

File: torchvision/transforms/transforms.py (hunk 1 above)
Comment (ekagra-ranjan) [reply torchvision/transforms/transforms.py]: What would you prefer for the dimension of mean_vector: 1xD or Dx1 given that transformation_matrix is DxD?

Comment (ekagra-ranjan) [reply torchvision/transforms/transforms.py]: According to the implementation of LinearTransformation I think it should be 1 x D.

I feel that the doc of the application of LinearTransformation about whitening is broken. The doc only shows the 1st line of the paragraph in the application block and the rest outside it. How do we write a multi-line bullet point? Do we replace `-` with `---`?

Comment (fmassa) [reply torchvision/transforms/transforms.py]: @ekagra-ranjan the dimension of the tensor doesn't matter, you can pass a `D` tensor and it will be fine, given that broadcasting will happen and the tensor will be filled with all zeros anyway.

About the doc, I'm not sure about how to properly format things. I'd try compiling the documentation locally and checking it visually

File: torchvision/transforms/transforms.py (hunk 3 above)
Comment (fmassa) [reply torchvision/transforms/transforms.py]: Oh yes, that's right, my bad!

File: torchvision/transforms/transforms.py [hunk 4]
```diff
@@ -737,17 +742,29 @@ def __call__(self, tensor):
...
 
+class LinearTransformation(AffineTransformation):
+    """
+    Note: This transform is deprecated in favor of AffineTransformation.
+    """
+
+    def __init__(self, transformation_matrix):
+        warnings.warn("The use of the transforms.LinearTransformation transform is deprecated, " +
+                      "please use transforms.AffineTransformation instead.")
+        super(LinearTransformation, self).__init__(transformation_matrix, torch.zeros(transformation_matrix.shape(0)).view(1, -1))
```
Comment (fmassa) [inline torchvision/transforms/transforms.py]: Can't you just make it as `torch.zeros_like(transformation_matrix[0])`?

File: torchvision/transforms/transforms.py [hunk 5]
```diff
@@ -738,17 +744,30 @@ def __call__(self, tensor):
...
 
+class LinearTransformation(AffineTransformation):
+    """
+    Note: This transform is deprecated in favor of AffineTransformation.
+    """
+
+    def __init__(self, transformation_matrix):
+        warnings.warn("The use of the transforms.LinearTransformation transform is deprecated, " +
+                      "please use transforms.AffineTransformation instead.")
+        super(LinearTransformation, self).__init__(transformation_matrix, torch.zeros(
```
Comment (fmassa) [inline torchvision/transforms/transforms.py]: This will fail if `transformation_matrix` is not of type `float32`, because by default `torch.zeros` returns a `float32` tensor. Can you use instead `zeros_like`? Also, why do you need to make it a `1 x D` tensor? Letting it be a `D` tensor is enough for broadcasting, and in my opinion nicer as well

Comment (ekagra-ranjan) [reply torchvision/transforms/transforms.py]: Done!

**Comment by ekagra-ranjan** [comment]
@fmassa  Is there anything else that you want me to do with this PR?

**Comment by ekagra-ranjan** [comment]
@fmassa Are the changes fine?

File: torchvision/transforms/transforms.py [hunk 6]
```diff
@@ -703,28 +703,34 @@ def __repr__(self):
...
+        mean_vector (Tensor): tensor [1 x D], D = C x H x W
     """
 
-    def __init__(self, transformation_matrix):
+    def __init__(self, transformation_matrix, mean_vector):
         if transformation_matrix.size(0) != transformation_matrix.size(1):
             raise ValueError("transformation_matrix should be square. Got " +
                              "[{} x {}] rectangular matrix.".format(*transformation_matrix.size()))
+
+        if mean_vector.size(1) != transformation_matrix.size(0):
```
Comment (fmassa) [inline torchvision/transforms/transforms.py]: this is raise an error if the user tries to use `LinearTransformation`, because the mean vector is 1d.

I really think that you should just make `mean_vector` here be 1d as well, and change it in the documentation.

Also, can you keep the tests for the `LinearTransformation`?

File: torchvision/transforms/transforms.py [hunk 7]
```diff
@@ -738,17 +744,29 @@ def __call__(self, tensor):
...
+        flat_tensor = tensor.view(1, -1) - self.mean_vector
         transformed_tensor = torch.mm(flat_tensor, self.transformation_matrix)
         tensor = transformed_tensor.view(tensor.size())
         return tensor
 
     def __repr__(self):
-        format_string = self.__class__.__name__ + '('
+        format_string = self.__class__.__name__ + '(transformation_matrix='
         format_string += (str(self.transformation_matrix.numpy().tolist()) + ')')
+        format_string += (", (mean_vector=" + str(self.mean_vector.numpy().tolist()) + ')')
```
Comment (fmassa) [inline torchvision/transforms/transforms.py]: no need to convert to numpy here, `tolist()` also exists for torch tensors

**Comment by yanfengliu** [comment]
Sorry if I misunderstood this, but I thought affine transformation means rotate, scale, shift, and shear, as shown in `torchvision.transforms.functional.affine(img, angle, translate, scale, shear, resample=0, fillcolor=None)`?

**Comment by fmassa** [comment]
Hum, there seems to be indeed a problem with our naming, which will indeed make this very confusing

Given that this is not in the 0.2.2 release, I'll rename this function with another name.
Any advices on the naming?

**Comment by ekagra-ranjan** [comment]
I have 2 suggestions for the name:
1) We can let the name remain LinearTransformation as the new transformation is still a linear operation ( (X-mean)/std  = AX - B, for some suitable A and B).

2) We can rename it as PixelNormalization since the transformation is basically is normalizing the values at each pixel in the image with pixel specific mean and standard deviation.

**Comment by fmassa** [comment]
@ekagra-ranjan I agree with leaving the name `LinearTransformation`.
Can you send a PR with this change?

**Comment by ekagra-ranjan** [comment]
Sure.
